File: SPCode/UI/Form/MasterForm.vb
Imports SoftLogik.Win.SPDataProxy
Imports System.ComponentModel

Namespace UI
    Public Class MasterForm

        Private _TypeID As String = String.Empty
        Private _NewRecordEmptyItem As DataRowView = Nothing

        Public Property TypeID() As String
            Get
                Return _TypeID
            End Get
            Set(ByVal value As String)
                _TypeID = value
            End Set
        End Property

        Protected Overridable Function NewMaster() As DataRowView
            Return SPMaster.GetEmptyMaster(TypeID).DefaultView.Item(0)
        End Function

        Protected Overrides Sub OnRecordChanged(ByVal e As SPFormRecordUpdateEventArgs)
            MyBase.OnRecordChanged(e)
            Dim recordRow As SPMasterRow = CType(e.DataRow, SPMasterRow)
            If recordRow IsNot Nothing Then
                Select Case e.DataState
                    Case SPFormDataStates.[New]
                        SPMaster.InsertMaster(recordRow.Name, recordRow.Note, TypeID)
                    Case SPFormDataStates.Edited
                        SPMaster.UpdateMaster(recordRow.MasterID, recordRow.Name, recordRow.Note)
                    Case SPFormDataStates.Deleted
                        SPMaster.DeleteMaster(recordRow.MasterID)
                End Select

            End If
        End Sub

        Protected Overrides Sub OnRecordBinding(ByVal e As SPFormRecordBindingEventArgs)
            MyBase.OnRecordBinding(e)
            If String.IsNullOrEmpty(TypeID) = False Then
                With e.BindingSettings
                    .DisplayMember = "Name"
                    .ValueMember = "MasterID"
                End With
                Try
                    e.DataSource = SPMaster.GetMaster(TypeID)
                    e.BindingSettings.NewRecordProc = AddressOf NewMaster
                Catch ex As Exception
                End Try
            End If


        End Sub


    End Class
End Namespace

File: SoftLogic.Win/UI/Form/MasterForm.cs
using System.Text.RegularExpressions;
using System.Diagnostics;
using System;
using System.Drawing.Drawing2D;
using System.Windows.Forms;
using System.Collections;
using System.Drawing;
using Microsoft.VisualBasic;
using System.Data.SqlClient;
using System.Data;
using System.Collections.Generic;
using WeifenLuo.WinFormsUI;
using Microsoft.Win32;
using WeifenLuo;
using System.ComponentModel;


namespace SoftLogik.Win.UI
	{
		public partial class MasterForm
		{
			public MasterForm()
			{
				InitializeComponent();
			}
			
			private string _TypeID = string.Empty;
			private DataRowView _NewRecordEmptyItem = null;
			
			public string TypeID
			{
				get
				{
					return _TypeID;
				}
				set
				{
					_TypeID = value;
				}
			}
			
			protected virtual DataRowView NewMaster()
			{
				return SPMaster.GetEmptyMaster(TypeID).DefaultView[0];
			}
			
			protected override void OnRecordChanged(SPFormRecordUpdateEventArgs e)
			{
				base.OnRecordChanged(e);
				SPMasterRow recordRow = (SPMasterRow) e.DataRow;
				if (recordRow != null)
				{
					switch (e.DataState)
					{
						case @SPFormDataStates.New:
							SPMaster.InsertMaster(recordRow.Name, recordRow.Note, TypeID);
							break;
						case SPFormDataStates.Edited:
							SPMaster.UpdateMaster(recordRow.MasterID, recordRow.Name, recordRow.Note);
							break;
						case SPFormDataStates.Deleted:
							SPMaster.DeleteMaster(recordRow.MasterID);
							break;
					}
					
				}
			}
			
			protected override void OnRecordBinding(SPFormRecordBindingEventArgs e)
			{
				base.OnRecordBinding(e);
				if (string.IsNullOrEmpty(TypeID) == false)
				{
					e.BindingSettings.DisplayMember = "Name";
					e.BindingSettings.ValueMember = "MasterID";
					try
					{
						e.DataSource = SPMaster.GetMaster(TypeID);
						e.BindingSettings.NewRecordProc = new System.EventHandler(NewMaster);
					}
					catch (Exception)
					{
					}
				}
				
				
			}
			
			
		}
	}